### PR TITLE
WIP: Ak/caching

### PR DESCRIPTION
**Pull request by mitochondrion**

### CODE REVIEW

File: TradeItIosTicketSDK2/TradeItLinkedBrokerManager.swift [hunk 1]
```diff
@@ -26,7 +26,7 @@ import PromiseKit
         let tradeItSession = tradeItSessionProvider.provide(connector: self.tradeItConnector)
         let linkedBroker = TradeItLinkedBroker(session: tradeItSession, linkedLogin: linkedLogin)
         
-        //we need to provide an error when we load from the keychain in order to the authenticate all method to handle this linkBroker
+        // TODO: Refactor the error code so we don't need to provide an error when we load from the keychain to indicate unauthenticated
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerManager.swift]: I actually don't think this needs to be refactored. It's kind of just like a dirty flag.

File: Podfile [hunk 2]
```diff
@@ -11,6 +11,7 @@ end
 
 def app_pods
   pod 'PromiseKit', '3.4'
+  pod 'SwiftyUserDefaults', '2.2.1'
```
Comment (jsom) [inline Podfile]: Add to podspec too

File: ExampleApp/ExampleViewController.swift [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import UIKit
-import TradeItIosTicketSDK2
+@testable import TradeItIosTicketSDK2
```
Comment (jsom) [inline ExampleApp/ExampleViewController.swift]: How come this is @testable here?

File: TradeItIosTicketSDK2/TradeItLinkedBrokerAccount.swift [hunk 4]
```diff
@@ -12,14 +15,12 @@ public class TradeItLinkedBrokerAccount: NSObject {
     var isEnabled = true
 
     init(linkedBroker: TradeItLinkedBroker,
-        brokerName: String,
          accountName: String,
          accountNumber: String,
-         balance: TradeItAccountOverview!,
-         fxBalance: TradeItFxAccountOverview!,
+         balance: TradeItAccountOverview?,
+         fxBalance: TradeItFxAccountOverview?,
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerAccount.swift]: 👍

File: TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    typealias SerializedLinkedBroker = [String: AnyObject]
+    typealias SerializedLinkedBrokers = [UserId: SerializedLinkedBroker]
+    typealias SerializedLinkedBrokerAccount = [String: String]
+    typealias SerializedLinkedBrokerAccounts = [AccountNumber: SerializedLinkedBrokerAccount]
+
+    private let ACCOUNTS_KEY = "ACCOUNTS"
+    private let ACCOUNTS_LAST_UPDATED_KEY = "ACCOUNTS_LAST_UPDATED"
+    private let ACCOUNT_NAME_KEY = "ACCOUNT_NAME"
+
+    var defaults = NSUserDefaults(suiteName: "it.trade")!
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift]: Should this be force unwrapped?

File: TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    func cache(linkedBroker linkedBroker: TradeItLinkedBroker) {
+        guard let userId = linkedBroker.linkedLogin.userId else { return }
+
+        var linkedBrokerCache = defaults[.linkedBrokerCache] as? SerializedLinkedBrokers ?? SerializedLinkedBrokers()
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift]: What is `SerializedLinkedBrokers` doing here? Creating an empty array of that type?

File: TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        linkedBrokerCache[userId] = serializedLinkedBroker
+
+        defaults[.linkedBrokerCache] = linkedBrokerCache
+    }
+
+    func syncFromCache(linkedBroker linkedBroker: TradeItLinkedBroker) {
+        guard let userId = linkedBroker.linkedLogin.userId
+            , let linkedBrokerCache = defaults[.linkedBrokerCache] as? SerializedLinkedBrokers
+            , let serializedLinkedBroker = linkedBrokerCache[userId] as SerializedLinkedBroker?
+        else { return }
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift]: Indent

File: TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    private func serialize(accounts accounts: [TradeItLinkedBrokerAccount]) -> SerializedLinkedBrokerAccounts {
+        var serializedAccounts = SerializedLinkedBrokerAccounts()
+
+        for account in accounts {
+            var serializedAccount = SerializedLinkedBrokerAccount()
+
+            serializedAccount[ACCOUNT_NAME_KEY] = account.accountName
+
+            serializedAccounts[account.accountNumber] = serializedAccount
+        }
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift]: Can you use `map` here instead?

File: TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+        for (accountNumber, serializedAccount) in serializedAccounts {
+            let account = TradeItLinkedBrokerAccount(linkedBroker: linkedBroker,
+                                                     accountName: serializedAccount[ACCOUNT_NAME_KEY] ?? "",
+                                                     accountNumber: accountNumber,
+                                                     balance: nil,
+                                                     fxBalance: nil,
+                                                     positions: [])
+            accounts.append(account)
+        }
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift]: Can you use `map` here instead?

File: TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+//        linkedBroker.accountsLastUpdated = serializedLinkedBroker[ACCOUNTS_LAST_UPDATED_KEY] ?? 0
+    }
+
+    func remove(linkedBroker linkedBroker: TradeItLinkedBroker) {
+        guard let userId = linkedBroker.linkedLogin.userId
+            , var linkedBrokerCache = defaults[.linkedBrokerCache] as? SerializedLinkedBrokers
+            else { return }
+
+        linkedBrokerCache[userId] = nil
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerCache.swift]: Is there something else you have to do here? Like persist the change back to NSUserDefaults?

File: TradeItIosTicketSDK2/TradeItLinkedBrokerManager.swift (hunk 1 above)
Comment (guillaumedebavelaere) [inline TradeItIosTicketSDK2/TradeItLinkedBrokerManager.swift]: Notice: this has been moved to the TradeItLinkedBroker Constructor

**Comment by jsom** [comment]
Merged already via `swift3` branch.

**Pull request closed without merging**